[1단계 방탈출 결제 / 배포] 포라(이승연) 미션 제출합니다.

nourzoo opened the pull request:
안녕하세요 베루스!
이번 미션의 [수정 범위](https://github.com/woowacourse/spring-roomescape-payment/pull/225/files/b5d0ec64b62a735ced64e3276a81a92786c1f8dd..45eecaf48c0ead003c89d850fd3e3973dce2df17)입니다!
잘 부탁드립니다 😊

## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr/studylogs/4450)에 셀프 체크를 작성했나요?

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

## 베이스 코드 선택 체크
- [x] 이전 미션의 내 코드에서 시작 
- [ ] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
### 현 상태

Spring은 레벨 2에서 처음 사용했고, 외부 API 연동과 http client 사용도 이번이 처음입니다.
RestClient 사용 방식이나 테스트 전략 등에서 부족한 부분이 많을 수 있습니다..😭

다양한 고민거리 주시면 감사하겠습니다! 🙏🏽

### 테스트 작성 시 겪은 고민

- **mock을 사용한 단위 테스트 한계**
    - RestClient 자체가 외부 API와 제대로 통신하는지 확인하기 어려웠습니다.
    - 결국 RestClient를 **직접 호출하는 테스트가 아니라**, 그 위에 얹힌 서비스 로직만 테스트하게 되었습니다.
- **현재 테스트 방식**
    1. 잘못된 `paymentKey`로 요청 시 실패하는지 확인
    2. 잘못된 `secretKey`로 요청 시 실패하는지 확인
    - 두 테스트는 각각 서로 다른 secretKey 값을 사용해야 했기 때문에, 테스트 환경 구성을 분리해서 진행했습니다. 하지만 이 테스트들은 **실제 외부 API에 의존**하기 때문에, 외부 API에 문제가 있을 경우 테스트가 실패할 가능성이 있습니다.
- **테스트에 대한 의문**
    - 외부 API를 직접 호출하면, `paymentKey`를 포함한 **결제 승인 요청 전체 플로우**에 대한 테스트는 어려웠습니다.  또한 테스트 환경에서는 `paymentKey`를 토스에게서 발급받을 수 없기 때문에 이 때 외부 API를 직접 호출하는 테스트 방식이 의미 있는가에 대한 의문이 들었습니다. 


### 해결 방법 고민

- **WireMock을 도입하여 RestClient 호출을 Stub 처리**하는 방법을 생각 중입니다
    - TossPaymentClient 내부에서 외부 API 호출 부분만 mock으로 대체합니다
    - 외부 의존성을 제거할 수 있습니다.
    - 다음 피드백에 반영해 볼 예정입니다!


### 궁금한 점

- RestClient 테스트는 어떤 방식이 가장 신뢰있을까요?
    - 외부 API를 직접 호출하는 방식은 재현할 수 있는 테스트가 한정적이고
    - mock을 사용하는 방식은 RestClient 전체 플로우 테스트가 어렵습니다.
    - WireMock 방식이 적절할까요? 혹은 더 나은 방식이 있을까요?


### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

nourzoo committed:
feat: setup project

jumdo12 committed:
docs: 1단계 README.md 추가

jumdo12 committed:
feat: 결제 기능 외부 서비스 이용하도록 기능 구현

jumdo12 committed:
feat: 외부 API 승인 기능 추가

nourzoo committed:
feat: Payment Service 추가, 결제 승인 실패 테스트 추가

jumdo12 committed:
refactor:PaymentService 를 Client 로 변경

jumdo12 committed:
test: Payment 테스트 작성

jumdo12 committed:
feat(Payment): 엔티티 추가

jumdo12 committed:
Test(Payment): 엔티티 테스트 추가

jumdo12 committed:
feat: Transaction 애노테이션 추가 및 OSIV 설정 false

jumdo12 committed:
fix: 테스트 데이터 날짜 수정

jumdo12 committed:
test: fake 테스트 삭제, 결제 승인 컨트롤러 테스트 오류 수정

jumdo12 committed:
fix: LocalDateTime 파싱 문제 개선

nourzoo committed:
feat: 타임아웃 설정 기능 추가

nourzoo committed:
feat: Payment 객체에 paymentKey 필드 추가

nourzoo committed:
feat: PaymentException에 상태코드 추가, LoginException -> UnauthorizedException으로 수정

nourzoo committed:
feat: 관리자 예약 생성 service 추가

nourzoo committed:
style: 불필요, 사용 안 하는 코드 삭제

verus-j commented on src/main/java/roomescape/common/config/TossPaymentClientConfig.java:
```diff
@@ -0,0 +1,28 @@
+public class TossPaymentClientConfig {
+    @Bean
+    public RestClient tossRestClient() {
+        String secretKey = "test_gsk_docs_OaPz8L5KdmQXkzRz3y47BMw6";
```
시크릿 키를 github에 올려도 될까요?

nourzoo replied:
시크릿 키는 서버 내부에서 관리해야 하며, 이 키를 통해 toss 서버와 소통하기 때문에 외부에 노출되면 안 됩니다.
그래서 application.yml에 선언하고, 이를 @Valid 어노테이션과 함께 필드 주입 방식으로 처리하도록 수정했습니다.

[feat: IllegalArgumentExcpetion 커스텀 예외 생성, @ControllerAdvice -> @RestC…](https://github.com/woowacourse/spring-roomescape-payment/pull/225/commits/98d42fb376c2140a8a32e2724037664b5858209f) 
[98d42fb](https://github.com/woowacourse/spring-roomescape-payment/pull/225/commits/98d42fb376c2140a8a32e2724037664b5858209f)

verus-j commented on src/main/java/roomescape/common/config/TossPaymentClientConfig.java:
```diff
@@ -0,0 +1,28 @@
+        String encodedAuth = "Basic " + Base64.getEncoder().encodeToString((secretKey + ":").getBytes(StandardCharsets.UTF_8));
+
+        SimpleClientHttpRequestFactory factory = new SimpleClientHttpRequestFactory();
+        factory.setConnectTimeout(3000);
+        factory.setReadTimeout(3000);
```
각각은 어떤 설정인가요? 그리고 3초로 세팅한 이유가 있나요?

nourzoo replied:
### connectTimeout 

클라이언트에서 설정한 시간까지 서버에 연결되지 않으면 발생합니다.
서버와 연결이 안 되는 상황에서 클라이언트가 계속 멈춰 있으면 안 되기 때문입니다.
예를 들어, 결제 요청 후 서버와 연결이 지연되면 사용자는 결제가 성공했는지 불안해할 수 있습니다.

### readTimeout

클라이언트에서 설정한 시간까지 서버에서 응답이 오지 않으면 발생합니다.
클라이언트와 서버는 연결이 됐지만, 서버가 클라이언트의 요청을 정상적으로 처리하지 못 할 때 발생합니다.
클라이언트에서 너무 많은 데이터를 조회했거나, 서버가 요청을 처리하는데 시간이 너무 오래 걸리는 경우 발생합니다.

결제와 같이 사용자가 민감한 기능에서는 기다리는 시간이 길어질수록 불안감이 커질 수 있습니다. 따라서 연결 지연이나 응답 지연이 발생할 때 빠르게 실패하도록 3초로 짧게 설정했습니다.

verus-j commented on src/main/java/roomescape/payment/client/TossPaymentClient.java:
```diff
@@ -0,0 +1,39 @@
+                            String errorBody = new String(res.getBody().readAllBytes(), StandardCharsets.UTF_8);
+                            TossErrorResponse errorResponse = objectMapper.readValue(errorBody, TossErrorResponse.class);
+                            throw new PaymentException("결제 실패 : " + errorResponse.message());
+                        }
```
메서드로 추출하면 어떨까요?

verus-j commented on src/main/java/roomescape/reservation/presentation/ReservationController.java:
```diff
@@ -46,9 +55,21 @@ public ResponseEntity<List<ReservationResponse>> getReservations(
+        TossPaymentResponse tossPaymentResponse = tossPaymentClient.confirmPayment(confirmRequest);
         ReservationResponse response = reservationService.createReservation(request, loginMember.id());
+        Payment payment = paymentService.save(tossPaymentResponse, response.id());
+
```
Reservation과 Payment를 생성하는건 같은 트랜잭션으로 묶을 필요가 없는걸까요?

nourzoo replied:
예약 생성은 결제 승인이 완료된 이후에 이루어지므로, 결제가 성공했다는 사실을 전제로 합니다. 따라서 예약을 확정 짓는 순간, 그에 대한 일종의 영수증인 Payment도 함께 저장되어야 합니다.
두 작업 중 하나라도 실패하면 전체 작업을 되돌려야 하므로 같은 트랜잭션 범위 내에서 관리하는 것이 맞는 것 같습니다!

ReservationService, PaymentService에서 각각 예약 확정, 영수증생성을 담당해서 이 두 서비스를 묶는 파서드 클래스를 생성하여 두 작업을 하나의 트랜잭션에서 처리 수 있도록 했습니다.

[feat: 예약 확정과 payment 생성 하나의 트랜잭션이 관리하도록 수정](https://github.com/woowacourse/spring-roomescape-payment/pull/225/commits/dcbd33fb190d404efd81bd92e551340a0c7bf14a)

verus-j commented on src/test/java/roomescape/payment/client/PaymentPaymentKeyTest.java:
```diff
@@ -0,0 +1,29 @@
+    TossPaymentClient tossPaymentClient;
+
+    @Test
+    void 클라이언트에서_획득하지_않은_페이먼트_키로_요청_시_실패한다() {
```
외부 API 연동과 관련된 테스트에 대해 본문에 고민을 적어두셨네요. 여기서 한번 얘기를 나눠보시죠!

여기 테스트에서는 tossPaymentClient를 실제 연동해서 테스트실행하셨군요. 매번 테스트 실행하마다 토스 서버는 정상이여야지만 테스트를 성공하겠네요. 간헐적으로 실패할 수 있는 테스트에 대해 포라는 어떻게 생각하시나요?

nourzoo replied:
외부 서버를 직접 부른다면 테스트가 외부 서버를 의존하게 됩니다. 예를 들어 토스 서버가 느리거나 장애가 발생하면, 코드가 완벽해도 테스트가 실패할 수 있기 때문에 신뢰성이 떨어집니다.

이 테스트는 `RestClient`을 사용하는 스프링 내부 단위테스트였고, 스프링이 아닌 다른 언어에서 재사용할 일이 없다고 생각했기 때문에 Spring이 제공하는 MockRestServiceServer를 쓰는 것이 더 효율적이라고 생각했습니다!

처음에는 `WireMock`을 도입해보려 `build.gradle`에 의존성도 추가해봤지만, 이 테스트의 목적에 비해서 설정과 작성에 드는 시간이 많이 든다고 판단했습니다. 테스트는 가볍고 빠르게 작성해야 한다고 생각해서 `MockRestServiceServer`로 리팩토링하게 되었습니다!

위의 이유에도 불구하고 `WireMock`을 만약 추천하신다면 그 이유가 궁금합니다!

[test: MockRestServiceServer 기반 토스 결제 API 테스트 추가](https://github.com/woowacourse/spring-roomescape-payment/pull/225/commits/447b558f8c5efcb4e3483f656b15d0eb30a5f5a1)

verus-j replied:
MockRestServiceServer를 사용해 테스트해도 문제가 없다고 생각합니다!

verus-j commented on README.md:
```diff
@@ -11,7 +11,7 @@
     - [x] 테마 조회 api 추가
 - [x] 사용자 관련 기능
```
제 로컬에서 테스트를 실행해보니 실패하는 테스트가 있네요 ㅠㅠ
확인부탁드려요~

verus-j reviewed (changes requested):
안녕하세요 포라
미션 진행하시느라 고생하셨습니다.
피드백 남겨뒀으니 확인부탁드려요!

nourzoo committed:
feat: IllegalArgumentExcpetion 커스텀 예외 생성, @ControllerAdvice -> @RestControllerAdvice로 변경 후 테스트 수정

nourzoo committed:
refactor: client용 패키지 분리

nourzoo committed:
fix: 테스트 오류 수정

nourzoo committed:
feat: 예약 확정과 payment 생성 하나의 트랜잭션이 관리하도록 수정

nourzoo committed:
test: MockRestServiceServer 기반 토스 결제 API 테스트 추가

nourzoo committed:
feat: ReservationPaymentFacade 빈등록 추가

nourzoo commented:
안녕하세요 베루스!
피드백 주신 부분들 반영하여 리뷰 남겼습니다.
이번 리뷰도 잘 부탁드리겠습니다 :) 😊

nourzoo committed:
test: MockRestServiceServer 이용한 client 테스트 리팩토링

nourzoo committed:
feat: Reservation에 상태 필드 추가, 결제 됐을 때 필드 바뀌도록 수정

nourzoo committed:
refactor: reservation 생성 메서드 분리

nourzoo committed:
feat: payment 객체에 nullable=false 추가

nourzoo committed:
feat: RestClient.Builder @Bean 분리

nourzoo committed:
fix: 테스트 코드 중 Reservation에 status 필드 추가

nourzoo committed:
fix: 결제 승인 테스트코드 추가

nourzoo committed:
feat: Reservation의 status 삭제, 결제 승인 후에 에러 핸들링 되도록 TossPaymentResponse, TossTossPaymentClient 수정

nourzoo committed:
fix: 결제 승인 실패 테스트 오류 수정

nourzoo committed:
fix: reservation 삭제 테스트 수정

nourzoo commented on src/main/java/roomescape/reservation/service/ReservationPaymentFacade.java:
```diff
@@ -0,0 +1,51 @@
+                request.amount(),
+                request.paymentKey()
+        );
+        ReservationWithPaymentResponse reservationWithPendingPayment = reservationService.createReservationWithPendingPayment(request, loginMember.id());
```
### ReservationStatus 필드 삭제

먼저 Reservation에 있던 `ReservationStatus`필드를 삭제했습니다.
예약 자체만으로는 취소된 예약을 db에 남길 필요가 없다고 생각했고, 오히려 결제 상태 여부는 payment에서 관리하는 것이 더 적절하다고 생각했습니다.

[토스 공식 문서 Payment 객체 설명](https://docs.tosspayments.com/reference#payment-%EA%B0%9D%EC%B2%B4
)에서도 Payment에 status로 결제의 상태를 식별할 수 있게 해두었습니다.

### 수정된 `결제 + 예약` 흐름

1. Reservation와 `PENDING` 상태의 Payment 객체 저장
2. toss 서버에 결제 승인 요청 - 트랜잭션X
3-1. 상태 코드가 2xx라면 Payment 상태를 `DONE`으로 변경
3-2. 상태 코드가 에러코드(4xx, 5xx)라면 Reservation 삭제, Payment 상태를 e로 변경
4. 에러 후처리

외부 API를 트랜잭션에서 분리하고, 에러가 발생하면 Payment 객체의 상태를 바꿀 수 있도록 하기 위해 위와같은 흐름이 나왔습니다. 지금과 같은 설계가 트랜잭션을 잘 이용한 설계인지 궁금합니다.

verus-j replied:
tossPaymentClient.confirmPayment(confirmRequest)에서 예외가 발생하면 PENDING 상태의 Payment 객체를 삭제할 필요가 없나요??

트랜잭션은 필요한 부분을 잘 분리하여 적용한 것 같습니다!

nourzoo commented on src/main/java/roomescape/client/TossPaymentClient.java:
```diff
@@ -0,0 +1,70 @@
+                    .body(request)
+                    .retrieve()
+                    .toEntity(TossPaymentResponse.class);
+        } catch (HttpStatusCodeException e) {
```
`RestClient`가 동기 방식이라 에러가 터지면 `retrieve().toEntity()` 내부에서 `HttpStatusCodeException`을 바로 던지게 됩니다.
이 때 json은 자동으로 역직렬화되지 않습니다.
저는 서비스 단에서 status를 기준으로 에러를 핸들링하고 싶었기 때문에, 예외를 잡아서 `TossPaymentResponse`객체로 역직렬화하도록 했습니다.

- 현재는 예외를 catch해서 직접 파싱해야 하는 구조인데, 이런 설계가 괜찮은지, 아니면 WebClient 같은 비동기 방식을 도입하여 응답이 성공이든 실패든 son 바디를 파싱하는게 나을지 궁금합니다!

verus-j replied:
상황을 보니 외부 서비스 코드에서 응답 바디 값을 필요한 상황이라 비동기로는 안될 것 같아 보이네요. onStatus와 같은걸 사용하면 안될려나요? 그리고 궁금한게 400 또는 500 대 에러가 나와도 항상 TossPaymentResponse로 변환이 가능한가요?

nourzoo reviewed (commented):
안녕하세요 베루스!
추가적으로 수정이 필요한 부분이 있어서 몇 가지 더 수정해 커밋 남겼습니다. 감사합니다! 😊

nourzoo committed:
feat: ReservationPaymentFacade 테스트 추가

- @SpringBootTest & @Transaction 사용해 db 롤백 처리
- 테스트 초기화 과정에서 EntityManager 사용
- TossPaymentClient 모킹 통해 외부 API 호출 테스트

nourzoo committed:
test: 필요없는 의존성 삭제

nourzoo committed:
feat: ReservationPaymentDataJpaFacade 테스트 추가

- @DataJpaTest 사용해 db 롤백 처리
- 테스트 초기화 과정에서 TestEntityManager 사용, Autowired로 주입
- TossPaymentClient 모킹 통해 외부 API 호출 테스트

nourzoo committed:
refactor: TestPropertySource로 sql 초기화 비활성화

nourzoo committed:
fix: 예약 대기 안 되는 버그 수정

nourzoo committed:
fix: ReservationTime 프록시 객체 비교 오류 버그 해결(getter 추가)

nourzoo committed:
refactor: 테스트에서 TestEntityManager 사용하도록 수정, application-test.properties 수정

verus-j reviewed (approved):
안녕하세요 포라
피드백 반영하고 코드 변경하신 것 확인했습니다.
다음 단계 진행하시면 됩니다~
